### DISCUSSION TITLE
React Native?

### DISCUSSION THREAD

**Discussion by andrioid** [automerge/automerge#1522] [General]
Is there any way of running this on platforms that don't support WASM yet?

This library would be a perfect fit for offline React Native applications and LLVM for persistence.

**Comment by alexjg** [comment]
I've laid out some of my thoughts on this [here](https://github.com/automerge/automerge/issues/525#issuecomment-1465168972) (with a follow up comment about react native specifically [here](https://github.com/automerge/automerge/issues/525#issuecomment-1466815900))

I would be extremely happy to see contributions related to this :smiley:

**Comment by andrioid** [comment]
https://github.com/cawfree/react-native-webassembly looks promising, and I hope I'll have time to try it out soon

**Comment by pwwolf** [comment]
Given this doesn't work in RN without presumably some hacking, is there any reason not to use the pure javascript implementation in RN? I was able to run automerge@1.0.1-preview.7 fine, but I was concerned there might be binary compatibility differences between that and another client running the wasm version.

**Comment by Tymek** [comment]
It's an interesting challenge. What I found so far:

### `window.WebAssembly` does not live in native land
Automerge Slim Next is using this global in `./wasm_bindgen_output/web/automerge_wasm.js`.
1. I doubt using above mentioned `react-native-webassembly` as global will just work. It's still worth a try.

There are other paths thru this integration hell. Options I found are:

2. Native module
https://github.com/dgca/expo-rust-demo/tree/main
Rust ➡️ Java+Swift+WASM ➡️ Native module
Most stable, but needs to be written 3 times, separately for each language.
From maintainer comment https://github.com/automerge/automerge/issues/525#issuecomment-1466815900 this is a preferred option.

3. [Turbo module](https://github.com/reactwg/react-native-new-architecture/blob/9845e0e0d709732a07c001684a7b11aa2e6eb1d5/docs/turbo-modules.md)
"Generate" native module from C++. Rust ➡️ Cpp ➡️ Native. Looks a bit round-about 🤷🏻‍♂️ 

4. [JSI module](https://blog.notesnook.com/getting-started-react-native-jsi/)
Write-once, but more experimental. See https://github.com/laptou/jsi-rs/tree/main

Similar issues where solved here: https://github.com/Nozbe/WatermelonDB/tree/master/native

**Comment by pvinis** [comment]
any of those would be great as a first version for react native/expo. get people hooked, then make a better v2 😅
